homeassistant/components/ws66i/media_player.py

```diff
@@ -0,0 +1,278 @@
+"""Support for interfacing with WS66i 6 zone home audio controller."""
+import logging
+
+from homeassistant import core
+from homeassistant.components.media_player import MediaPlayerEntity
+from homeassistant.components.media_player.const import (
+    SUPPORT_SELECT_SOURCE,
+    SUPPORT_TURN_OFF,
+    SUPPORT_TURN_ON,
+    SUPPORT_VOLUME_MUTE,
+    SUPPORT_VOLUME_SET,
+    SUPPORT_VOLUME_STEP,
+)
+from homeassistant.const import (
+    CONF_IP_ADDRESS,
+    EVENT_HOMEASSISTANT_STOP,
+    STATE_OFF,
+    STATE_ON,
+)
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import config_validation as cv, entity_platform, service
+
+from .const import (
+    CONF_SOURCES,
+    DOMAIN,
+    FIRST_RUN,
+    SERVICE_RESTORE,
+    SERVICE_SNAPSHOT,
+    WS66I_OBJECT,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+PARALLEL_UPDATES = 1
+
+SUPPORT_WS66I = (
+    SUPPORT_VOLUME_MUTE
+    | SUPPORT_VOLUME_SET
+    | SUPPORT_VOLUME_STEP
+    | SUPPORT_TURN_ON
+    | SUPPORT_TURN_OFF
+    | SUPPORT_SELECT_SOURCE
+)
+
+
+@core.callback
+def _get_sources_from_dict(data):
+    sources_config = data[CONF_SOURCES]
+
+    source_id_name = {int(index): name for index, name in sources_config.items()}
+
+    source_name_id = {v: k for k, v in source_id_name.items()}
+
+    source_names = sorted(source_name_id.keys(), key=lambda v: source_name_id[v])
+
+    return [source_id_name, source_name_id, source_names]
+
+
+@core.callback
+def _get_sources(config_entry):
+    if CONF_SOURCES in config_entry.options:
+        data = config_entry.options
+    else:
+        data = config_entry.data
+    return _get_sources_from_dict(data)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the WS66i 6-zone amplifier platform."""
+    addr = config_entry.data[CONF_IP_ADDRESS]
+
+    # Open up a connection with the WS66i
+    ws66i = hass.data[DOMAIN][config_entry.entry_id][WS66I_OBJECT]
+    try:
+        await hass.async_add_executor_job(ws66i.open)
+    except ConnectionError as err:
+        raise ConfigEntryNotReady from err
+
```

Comment:
Do this in `__init__`'s `async_setup_entry` so you don't get to platform setup if there is a ConnectionError

Reply:
ws66i.open() creates a connection with the amplifier. I call this here and keep it open until the component is unloaded. However, I also call ws66i.open in my __init__'s async_setup_entry as well to verify that a connection can be made, and then I close it.

Reply:
This needs to be done in `__init__` as `ConfigEntryNotReady` does not work in a platform